src/Opaleye/Internal/HaskellDB/PrimQuery.hs

```diff
@@ -60,6 +60,7 @@ data BinOp      = (:==) | (:<) | (:<=) | (:>) | (:>=) | (:<>)
                 | (:->) | (:->>) | (:#>) | (:#>>)
                 | (:@>) | (:<@) | (:?) | (:?|) | (:?&)
                 | (:&&) | (:<<) | (:>>) | (:&<) | (:&>) | (:-|-)
+                | OpArrayIndex
```

Comment:
I'm not sure the historical reasons that all `BinOp`s are infix, but since they are it seems neatest to keep it that way for now and add a separate `ArrayIndex` constructor.